src/group_theory/double_coset.lean

```diff
@@ -0,0 +1,331 @@
+this is the usual left or right quotient of a group by a subgroup.
+
+## Main definitions
+
+* `double_coset_rel`: The double coset relation defined by two subgroups `H K` of `G`.
+* `double_coset.quotient`: The quotient of `G` by the double coset relation, i.e, ``H \ G / K`.
+-/
+
+
+variables {G : Type*} [group G] {α : Type*} [has_mul α] (J: subgroup G) (g : G)
+
+namespace double_coset
+
+open_locale pointwise
+
+/--The double_coset as an element of `set α` corresponding to `s a t` -/
+def doset (a : α) (s t : set α) : set α := s * {a} * t
+
+@[simp]
+lemma doset_mem (s t : set α) (a b : α) : b ∈ (doset a s t) ↔ ∃ x : (s × t), b = x.1 * a * x.2 :=
+⟨λ ⟨_, y, ⟨x, _, hx, rfl, rfl⟩, hy, h⟩, ⟨⟨⟨x, hx⟩, ⟨y, hy⟩⟩, h.symm⟩,
+  λ ⟨⟨x, y⟩, h⟩, ⟨x * a, y, ⟨x, a, x.2, rfl, rfl⟩, y.2, h.symm⟩⟩
+
+def double_coset_rel (H K : set G) : G → G → Prop :=
+λ x y, doset x H K = doset y H K
+
+lemma rel_reflex {H K : set G} : reflexive (double_coset_rel H K) :=
+λ x, rfl
+
+lemma rel_symm {H K : set G} : symmetric (double_coset_rel H K) :=
+λ x y, eq.symm
+
+lemma rel_trans {H K : set G} : transitive (double_coset_rel H K) :=
+λ x y z, eq.trans
+
+lemma rel_is_equiv {H K : set G} : equivalence (double_coset_rel H K) :=
+⟨rel_reflex, rel_symm, rel_trans⟩
+
+/-- The setoid defined by the double_coset relation -/
+def setoid (H K : set G) : setoid G :=
+⟨double_coset_rel H K, rel_is_equiv⟩
+
+/-- Quotient of `G` by the double coset relation, i.e. `H \ G / K` -/
+def quotient (H K : set G) : Type* :=
+quotient (setoid H K)
+
+lemma doset_subgroups_mem_self (H K : subgroup G) (a : G) : a ∈ (doset a H K) :=
+begin
+rw doset_mem,
+use 1,
+simp only [mul_one, one_mul, prod.snd_one, subgroup.coe_one, prod.fst_one],
+end
+
+lemma sub_doset  (H K : subgroup G) (a b : G) (hb : b ∈ doset a H K) :
+  (doset b H.1 K ) = (doset a H K ) :=
+begin
+  obtain ⟨_, k, ⟨h, a, hh, (rfl : _ = _), rfl⟩, hk, rfl⟩ := hb,
+  rw [doset, doset, ←set.singleton_mul_singleton, ←set.singleton_mul_singleton],
+  rw [mul_assoc, mul_assoc, ←mul_assoc, ←mul_assoc H.carrier],
+  have key1 : H.carrier * {h} = H :=
+  by {simp only [set.image_mul_right, subgroup.mem_carrier,
+  set_like.mem_coe, set.mul_singleton] at *,
+  ext1,
+  simp only [set.mem_preimage, subgroup.mem_carrier, set_like.mem_coe] at *,
+  fsplit,
+  intro hxh,
+  have:= subgroup.mul_mem H hxh hh,
+  simp only [inv_mul_cancel_right] at this,
+  apply this,
+  intro hx,
+  have hinv:= subgroup.inv_mem H hh,
+  apply  subgroup.mul_mem H hx hinv,},
+  have key2 : ({k} : set G) * K = K :=
+  by{simp only [set.image_mul_left, set.singleton_mul],
+  ext1,
+  split,
+  simp only [set.mem_preimage, set_like.mem_coe],
+  intro hkx,
+  have:= subgroup.mul_mem K hk hkx,
+  simp only [mul_inv_cancel_left] at this,
+  apply this,
+  intro hx,
+  simp only [set.mem_preimage, set_like.mem_coe],
+  have hinv:= subgroup.inv_mem K hk,
+  apply subgroup.mul_mem K hinv hx,},
+  rw [key1, key2],
+end
```

Comment:
The two lemmas I extracted should be moved to another file.
```suggestion
lemma _root_.subgroup_mul_singleton {H : subgroup G} {h : G} (hh : h ∈ H) :
  (H : set G) * {h} = H :=
begin
  refine le_antisymm _ (λ h' hh',
    ⟨h' * h⁻¹, h, H.mul_mem hh' (H.inv_mem hh), rfl, inv_mul_cancel_right h' h⟩),
  rintros _ ⟨h', h, hh', rfl : _ = _, rfl⟩,
  exact H.mul_mem hh' hh,
end

lemma _root_.singleton_mul_subgroup {H : subgroup G} {h : G} (hh : h ∈ H) :
  {h} * (H : set G) = H :=
begin
  refine le_antisymm _ (λ h' hh', ⟨h, h⁻¹ * h', rfl, H.mul_mem (H.inv_mem hh) hh',
    mul_inv_cancel_left h h'⟩),
  rintros _ ⟨h, h', rfl : _ = _, hh', rfl⟩,
  exact H.mul_mem hh hh',
end

lemma sub_doset {H K : subgroup G} {a b : G} (hb : b ∈ doset a H K) :
  doset b H K = doset a H K :=
begin
  obtain ⟨_, k, ⟨h, a, hh, (rfl : _ = _), rfl⟩, hk, rfl⟩ := hb,
  rw [doset, doset, ←set.singleton_mul_singleton, ←set.singleton_mul_singleton, mul_assoc,
      mul_assoc, singleton_mul_subgroup hk, ←mul_assoc, ←mul_assoc, subgroup_mul_singleton hh],
end
```

Reply:
Do you have any preference for where to put it? I thought it might well go in `group_theory/subgroup/pointwise` but I'm not sure if there is a better place.

Reply:
Another option is `algebra/pointwise`, since `group_theory/subgroup/pointwise` seems more focused on `smul` as opposed to `mul`.

Reply:
So I think the file containing the definition of subgroup depends on (at some level) `algebra/pointwise`. So I might just put it in `group_theory/subgroup/pointwise`